lib/ts/recipe/totp/recipe.ts

```diff
@@ -0,0 +1,227 @@
+import RecipeModule from "../../recipeModule";
+import RecipeImplementation from "./recipeImplementation";
+import APIImplementation from "./api/implementation";
+import { APIHandled, HTTPMethod, NormalisedAppinfo, RecipeListFunction } from "../../types";
+import { CREATE_DEVICE_API, VERIFY_CODE_API, VERIFY_DEVICE_API, LIST_DEVICE_API, REMOVE_DEVICE_API } from "./constants";
+import {
+    TypeInput,
+    TypeNormalisedInput,
+    RecipeInterface,
+    APIInterface,
+    GetUserIdentifierInfoForUserIdFunc,
+} from "./types";
+import { validateAndNormaliseUserInput } from "./utils";
+import STError from "./error";
+import { Querier } from "../../querier";
+import { BaseRequest, BaseResponse } from "../../framework";
+import createDeviceAPI from "./api/createDevice";
+import verifyDeviceAPI from "./api/verifyDevice";
+import verifyCodeAPI from "./api/verifyCode";
+import removeDeviceAPI from "./api/removeDevice";
+import listDevicesAPI from "./api/listDevices";
+import { sendNon200ResponseWithMessage } from "../../utils";
+import TotpError from "./error";
+import { getUser } from "../..";
+
+export default class TotpRecipe extends RecipeModule {
+    private static instance: TotpRecipe | undefined = undefined;
+    static RECIPE_ID = "totp";
+
+    config: TypeNormalisedInput;
+
+    recipeInterfaceImpl: RecipeInterface;
+
+    apiImpl: APIInterface;
+
+    isInServerlessEnv: boolean;
+
+    constructor(recipeId: string, appInfo: NormalisedAppinfo, isInServerlessEnv: boolean, config: TypeInput) {
+        super(recipeId, appInfo);
+        this.isInServerlessEnv = isInServerlessEnv;
+        this.config = validateAndNormaliseUserInput(this, appInfo, config);
+
+        {
+            let builder = new OverrideableBuilder(
+                RecipeImplementation(Querier.getNewInstanceOrThrowError(recipeId), this.config)
+            );
+            this.recipeInterfaceImpl = builder.override(this.config.override.functions).build();
+        }
+        {
+            let builder = new OverrideableBuilder(APIImplementation());
+            this.apiImpl = builder.override(this.config.override.apis).build();
+        }
+    }
+
+    static getInstanceOrThrowError(): TotpRecipe {
+        if (TotpRecipe.instance !== undefined) {
+            return TotpRecipe.instance;
+        }
+        throw new Error("Initialisation not done. Did you forget to call the SuperTokens.init or totp.init function?");
+    }
+
+    static init(config?: TypeInput): RecipeListFunction {
+        return (appInfo, isInServerlessEnv) => {
+            if (TotpRecipe.instance === undefined) {
+                TotpRecipe.instance = new TotpRecipe(TotpRecipe.RECIPE_ID, appInfo, isInServerlessEnv, config ?? {});
+                return TotpRecipe.instance;
+            } else {
+                throw new Error("TOTP recipe has already been initialised. Please check your code for bugs.");
+            }
+        };
+    }
+
+    static reset() {
```

Comment:
you need to call this function in the test folder - in utils.js

Reply:
done